workflows/metatdenovo.nf

```diff
@@ -73,17 +144,373 @@ workflow METATDENOVO {
+        UNPIGZ_GFF(PROKKA_SUBSETS.out.gff.map { meta, gff -> [ [id: "${params.orf_caller}.${meta}"], gff ] })
+        ch_versions      = ch_versions.mix(PROKKA_SUBSETS.out.versions)
+        ch_gff           = UNPIGZ_GFF.out.unzipped
+        ch_protein       = PROKKA_SUBSETS.out.faa
+        ch_multiqc_files = ch_multiqc_files.mix(PROKKA_SUBSETS.out.prokka_log.collect{ meta, log -> log } )
+     }
+
+    //
+    // MODULE: Run PRODIGAL on assembly output.
+    //
+    if ( orf_caller == 'prodigal' ) {
+        PRODIGAL( ch_assembly_contigs.map { meta, contigs -> [ [id: "${assembler}.${orf_caller}"], contigs  ] } )
+        UNPIGZ_GFF(PRODIGAL.out.gff.map { meta, gff -> [ [id: "${meta.id}.${orf_caller}"], gff ] })
+        ch_gff          = UNPIGZ_GFF.out.unzipped
+        ch_protein      = PRODIGAL.out.faa
+        ch_versions     = ch_versions.mix(PRODIGAL.out.versions)
+    }
+
+    //
+    // SUBWORKFLOW: run TRANSDECODER. Orf caller alternative for eukaryotes.
+    //
+    if ( orf_caller == 'transdecoder' ) {
+        TRANSDECODER ( ch_assembly_contigs.map { meta, contigs -> [ [id: "transdecoder.${meta.id}" ], contigs ] } )
+        ch_gff      = TRANSDECODER.out.gff
+        ch_protein  = TRANSDECODER.out.pep
+        ch_versions = ch_versions.mix(TRANSDECODER.out.versions)
+    }
+
+    // Populate channels if the user provided the orfs
+    if ( orf_caller == 'user_orfs' ) {
+        Channel
+            .value ( [ [ id: "${assembler}.${orf_caller}" ], file(params.gff) ] )
+            .set { ch_gff }
+        Channel
+            .value ( [ [ id: "${assembler}.${orf_caller}" ], file(params.protein_fasta) ] )
+            .set { ch_protein }
+    }
+
+    //
+    // MODULE: Create a BBMap index
+    //
+    BBMAP_INDEX(ch_assembly_contigs.map { meta, contigs -> contigs })
+    ch_versions   = ch_versions.mix(BBMAP_INDEX.out.versions)
+
+    //
+    // MODULE: Call BBMap with the index once per sample
+    //
+    BBMAP_ALIGN ( ch_clean_reads, BBMAP_INDEX.out.index )
+    ch_versions = ch_versions.mix(BBMAP_ALIGN.out.versions)
+
+    //
+    // SUBWORKFLOW: classify ORFs with a set of hmm files
+    //
+    ch_hmmrs
+        .combine(ch_protein)
+        .map { hmm, meta, protein ->[ [ id: "${assembler}.${orf_caller}" ], hmm, protein ] }
+        .set { ch_hmmclassify }
+    HMMCLASSIFY ( ch_hmmclassify )
+    ch_versions = ch_versions.mix(HMMCLASSIFY.out.versions)
+
+    //
+    // MODULE: FeatureCounts. Create a table for each samples that provides raw counts as result of the alignment.
+    //
+
+    BAM_SORT_STATS_SAMTOOLS ( BBMAP_ALIGN.out.bam, ch_assembly_contigs )
+    ch_versions = ch_versions.mix(BAM_SORT_STATS_SAMTOOLS.out.versions)
+
+    BAM_SORT_STATS_SAMTOOLS.out.bam
+        .combine(ch_gff.map { meta, bam -> bam } )
+        .set { ch_featurecounts }
+
+    ch_collect_stats
+        .combine(BAM_SORT_STATS_SAMTOOLS.out.idxstats.collect { meta, idxstats -> idxstats }.map { [ it ] } )
```

Comment:
What's going on here with the `.map { [ it ] }`? It's already a list.

Reply:
If I will not use map, I will get this form of the list:
`[/Users/dadiaa/projects/methods/nf-core-dev/metatdenovo/work/b5/e4f760099cf880cf5429db68259e0a/SAMPLE3_PE.idxstats, /Users/dadiaa/projects/methods/nf-core-dev/metatdenovo/work/3d/2b60c582dfdc3481838fbc764fa64a/SAMPLE2_PE.idxstats, /Users/dadiaa/projects/methods/nf-core-dev/metatdenovo/work/ca/94c0f1f9aa43f288e00c7531f6cd83/SAMPLE1_PE.idxstats]`

but I would like this form:
`[[/Users/dadiaa/projects/methods/nf-core-dev/metatdenovo/work/ca/94c0f1f9aa43f288e00c7531f6cd83/SAMPLE1_PE.idxstats, /Users/dadiaa/projects/methods/nf-core-dev/metatdenovo/work/3d/2b60c582dfdc3481838fbc764fa64a/SAMPLE2_PE.idxstats, /Users/dadiaa/projects/methods/nf-core-dev/metatdenovo/work/b5/e4f760099cf880cf5429db68259e0a/SAMPLE3_PE.idxstats]]`

Reply:
Why do you need the second form?

Reply:
I don't fully understand why, but it works in particular for `ch_collect_stats`

Reply:
Ah, I understand now. If one doesn't do it you'll get `[ ch_collect_statsJ, idxstats, idxstats2, ..., idxstatsN ]` when you really need `[ ch_collect_statsJ, [idxstats1, idxstats2, ..., idxstatsN] ]`

Reply:
Might be worth reporting this as a bug. In the case when no `by` is used, there should just be a straight cartesian product regardless of what Object type it is.

Reply:
OK, so I will not change this part and the others with the same structure! I can write you in private to talk about the bug report, I would like to understand better how I can formulate the bug, you know much better than me about it!